File: Python3/heaters.py
class Solution:
    def findRadius(self, houses: List[int], heaters: List[int]) -> int:
        houses = sorted(houses)
        heaters = sorted(heaters)
        
        if len(heaters) == 1:
            return max(heaters[0] - houses[0], houses[-1] - heaters[0])
        elif len(houses) == 1:
            return min([abs(houses[0]-heaters[i]) for i in range(len(heaters))])
        else:
            radius = 0
            pos = 0
            heaters = [float('-inf')] + heaters + [float('inf')]
            
            for house in houses:
                while house >= heaters[pos]:
                    pos += 1
                tmp = min(house - heaters[pos - 1], heaters[pos] - house)
                radius = max(radius, tmp)
                
            return radius

File: C/heaters.c
int compare(const void* a, const void* b){
    return *(int*)a - *(int*)b;
}


int findRadius(int* houses, int housesSize, int* heaters, int heatersSize){ 
    qsort(houses, housesSize, sizeof(int), compare);
    qsort(heaters, heatersSize, sizeof(int), compare);
        
    if(heatersSize == 1)
        return fmax(heaters[0]-houses[0], houses[housesSize-1]-heaters[0]);
    else {
        int radius = 0;
        int pos = 0;
           
        for(int i = 0; i < housesSize; i++) {
            int cur = houses[i];
                
            while(pos < heatersSize - 1 && abs(heaters[pos + 1] - cur) <= abs(heaters[pos] - cur))                     
                pos ++;
                
            radius = fmax(radius, abs(heaters[pos] - cur));
        }
        
        return radius;
    }
}
